### ISSUE TITLE
Scale for axis with string values

### ISSUE THREAD

**Issue by arogozhnikov** [altair-viz/altair#2680]
I'm trying to use 'lookup scaling' (don't know the right term) for axis encoding the same way it works for other encodings (e.g. for colors and sizes), but it fails at vega-validation

```python
import pandas as pd
import altair as alt

df = pd.DataFrame(dict(
    x=['A', 'B', 'C', 'D'], 
    y=[1, 2, 3, 4]
))

alt.Chart(df).mark_point().encode(
    x=alt.X('x', 
        scale=alt.Scale(
            domain=['A', 'B', 'C', 'D'],
            range=[1, 2, 11, 12],
        ),
    ),
    y=alt.Y('y'),
)
```

Result is a JS error:
```
Javascript Error: layout size is step although width/height is not step.
This usually means there's a typo in your chart specification. See the javascript console for the full traceback.
```

If I provide height and width explicitly, resulting values are treated as pixels (so 800 are total pixels in the example below and points are put at 101, 102, 211, 212), and I found no way to change this behavior 
```
alt.Chart(df).mark_point().encode(
    x=alt.X('x', 
        scale=alt.Scale(
            domain=['A', 'B', 'C', 'D'],
            range=[100, 102, 211, 212],
        ),
    ),
    y=alt.Y('y'),
).properties(
    width=800,
    height=100
)
```

**Comment by joelostblom** [comment]
I don't think you can apply a numerical `range` while setting the `domain` to string labels. Could you explain what you expected output would be from that code? You can also read more about the `range` parameter in the VegaLite documentation https://vega.github.io/vega-lite/docs/scale.html#range

**Comment by arogozhnikov** [comment]
> I don't think you can apply a numerical range while setting the domain to string labels

you're right... only first two items of list are taken into account and interpreted as min/max for range. Bummer

> Could you explain what you expected output would be from that code?

Sure. Maybe I'll start with a bit of motivation? I miss multi-level axes in altair very much. In different variations these plots are quite common in my work. How this may look like (could be a bar chart too with errors and points as overlay):
![Screen Shot 2022-09-19 at 8 17 31 PM](https://user-images.githubusercontent.com/6318811/191160931-b72657e4-1df2-48b1-a325-76ee963be632.png)

There may be one level or two levels or three levels in the axis (so here two levels - treatment and amount)
Now, yes, I am aware about stacking charts (and Grouped Bar chart example in documentation), but neither combination provided a coherent result.

So I though maybe if I can use range to map every value of discrete axis to a position I can create small visual gaps between groups. No idea how to align second-third level (treatments in this example), but already an improvement

**Comment by joelostblom** [comment]
Thanks for elaborating! I noticed that you said that the grouped bar example in the docs does not produce a "coherent result", but I am not sure I quite understand what you mean with that. With a few modifications to the docs example I can get something that is looks quite similar to your example:

```py
import altair as alt
from vega_datasets import data

source = data.barley()

alt.Chart(source).mark_point().encode(
    y=alt.Y('year:O', title=''),
    x='sum(yield):Q',
    color='year:N',
    row=alt.Row('site:N', header=alt.Header(labelAngle=0, labelAlign='left'), title='')
)
```

![image](https://user-images.githubusercontent.com/4560057/191566530-ac351e3b-08a6-4b48-88b1-96784f1216ab.png)

**Comment by arogozhnikov** [comment]
First issue here is that y axis in all plots in your example is the same. Try with this source instead and you'll see what I mean:
```
source = data.barley()
source['year'] += source['site'].map(len)
```
Second issue is no simple way to have three levels of hierarchy like culture-city-year instead of city-year

**Comment by joelostblom** [comment]
1. If you are referring to the y-axis being different for each plot you can solve that with `resolve_scale`.
2. I only saw two levels in your example above, but you're right that you can't have three levels of **labels** on the axis. You can however use the new `x/yOffset` encoding together with color, to create three levels of separation along the axis.

This example includes both the points above:

```py
import altair as alt
from vega_datasets import data

source = data.barley().query('variety == ["Velvet", "Trebi"]')
source['year'] += source['site'].map(len)

alt.Chart(source).mark_point().encode(
    x='sum(yield):Q',
    y=alt.Y('year:O', title=''),
    yOffset='variety',
    color='variety',
    row=alt.Row('site:N', header=alt.Header(labelAngle=0, labelAlign='left'), title='')
).resolve_scale(
    y='independent'
)
```

![image](https://user-images.githubusercontent.com/4560057/191580276-44784ac3-4b77-49ff-b1aa-19a350a635a3.png)

**Comment by arogozhnikov** [comment]
that's neat, thanks for your ideas. yOffset looks very interesting.

Unfortunately, it can't be assumed that variety is the same or similar for every location - conditions at the best form a tree . 

We've got too far from the original discussion here, original question was largely my misunderstanding of scale capabilities, let me close the issue. 

If you consider suggestions, I think that hierarchy axes are first-class citizens. vega seems to think differently about it.

I've implemented a simple, compact and not very attractive version that allows me to move on:
![visualization (3)](https://user-images.githubusercontent.com/6318811/191592196-b8c347c9-63ec-4c37-9010-b1b5c9329b30.png)
